vault/plugin_catalog.go

```diff
@@ -333,43 +384,85 @@ func (c *PluginCatalog) getPluginTypeFromUnknown(ctx context.Context, logger log
 	}
 	merr = multierror.Append(merr, err)
 
+	pluginType, err := c.getBackendPluginType(ctx, plugin)
+	if err == nil {
+		return pluginType, nil
+	}
+	merr = multierror.Append(merr, err)
+
+	return consts.PluginTypeUnknown, merr
+}
+
+// getBackendPluginType returns an error if the plugin is not a backend plugin.
+func (c *PluginCatalog) getBackendPluginType(ctx context.Context, pluginRunner *pluginutil.PluginRunner) (consts.PluginType, error) {
+	merr := &multierror.Error{}
 	// Attempt to run as backend plugin
-	client, err := backendplugin.NewPluginClient(ctx, nil, plugin, log.NewNullLogger(), true)
+	config := pluginutil.PluginClientConfig{
+		Name:            pluginRunner.Name,
+		PluginSets:      backendplugin.PluginSet,
+		HandshakeConfig: backendplugin.HandshakeConfig,
+		Logger:          log.NewNullLogger(),
+		IsMetadataMode:  false,
+		AutoMTLS:        true,
+	}
+
+	var client logical.Backend
+	// Attempt to run as backend V5 plugin
+	c.logger.Debug("attempting to load credential backend plugin", "name", pluginRunner.Name)
+	pc, err := c.newPluginClient(ctx, pluginRunner, config)
 	if err == nil {
-		err := client.Setup(ctx, &logical.BackendConfig{})
+		// dispense the plugin so we can get its type
+		client, err = backendplugin.Dispense(pc.ClientProtocol, pc)
+		if err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("failed to load plugin as backend v5: %w", err))
```

Comment:
If we enter this block where err != nil, we'll next go to line 463 (`client.Setup()`). Is it possible that client could be nil at that point? I was thinking we'd try to dispense the v4 plugin in this case, but the if/else logic isn't structured that way. I could be missing something though 🤔

Reply:
Good questions!

> Is it possible that client could be nil at that point?

This _should_ be a very rare case. I was able to get Dispense to fail by misconfiguring the [PluginSet](https://github.com/hashicorp/vault/blob/4d7dbd1c90ecb2d3e19caf555a7ad707ee64d19f/sdk/plugin/plugin_v5.go#L94). That being said, we should guard against it.

> I was thinking we'd try to dispense the v4 plugin in this case, but the if/else logic isn't structured that way

I think you are right that we should attempt to fallback here. If Dispense fails in this code path we fail registration. We actually _do_ fallback to v4 in the case where the plugin type is provided to the "register" command so I think we should do the same here.

I will update this, thanks!